### PR TITLE
OCD-3752 - Convert Test Procedures to Conformance Methods

### PR DESCRIPTION
**Pull request by tmy1313**

### CODE REVIEW

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/onetime/conformancemethodconversion/ConformanceMethodConversionJob.java [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    }
+
+    private Boolean doesTestProcedureMatchRule(CertificationResultTestProcedure crtp, Boolean hasGap, ConversionRule rule) {
+        Boolean gapCheck = rule.getHasGap() == null ? true : hasGap == rule.getHasGap();
+        return (StringUtils.isNotEmpty(crtp.getTestProcedure().getName())
+                        && rule.getTestProcedureName().equals(WILDCARD)
+                        && gapCheck)
+                || (StringUtils.isNotEmpty(crtp.getTestProcedure().getName())
+                        && rule.getTestProcedureName().equals(crtp.getTestProcedure().getName())
+                        && gapCheck);
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/onetime/conformancemethodconversion/ConformanceMethodConversionJob.java]: I feel like this statement is really hard to parse. Both of the main clauses on each side of the `||` have two of the three sub-clauses that are identical. It's like

```
if (a & b1 & c 
|| a & b2 & c)
```

Would it be the same logic / read more clearly if it were instead something like
```
if (a & c & (b1 | b2))
```

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/onetime/conformancemethodconversion/ConformanceMethodConversionJob.java [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+            // https://stackoverflow.com/questions/3037006/starting-new-transaction-in-spring-bean
+            TransactionTemplate txTemplate = new TransactionTemplate(txManager);
+            txTemplate.setPropagationBehavior(TransactionDefinition.PROPAGATION_REQUIRES_NEW);
+            txTemplate.execute(new TransactionCallbackWithoutResult() {
+
+                @Override
+                protected void doInTransactionWithoutResult(TransactionStatus status) {
+                    try {
+
+                        getAll2015CertifiedProducts(LOGGER, 2).stream()
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/onetime/conformancemethodconversion/ConformanceMethodConversionJob.java]: I have some concern extending the CertifiedProduct2015Gatherer and using this method which retrieves the 2015 listing details from the cache. The concern is due to some issues I ran into with the merge and split of developers not recording activity (the jobs were getting the "before" listing and "after" listing from the cache, so not registering the change). If this job were guaranteed to only run immediately following the cache being populated by the overnight jobs that would alleviate my concern. If it is going to be run at any other time then I am not sure about using the cached listing details since the test procedures for that listing could potentially have been changed between cache population and job run. I completely realize this is unlikely but the cached data burned me twice for that recent ticket so I am sensitive to it.

**Review by kekey1 — DISMISSED** [review]

**Review by andlar — DISMISSED** [review]

**Review by andlar — DISMISSED** [review]

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/CertifiedProduct2015Gatherer.java [hunk 3]
```diff
@@ -24,25 +24,42 @@ public class CertifiedProduct2015Gatherer {
...
+    public List<CertifiedProductSearchDetails> getAll2015CertifiedProducts(Logger logger, Boolean useListingCache) {
         logger.info("Retrieving all 2015 listings");
         List<CertifiedProductDetailsDTO> listings = certifiedProductDAO.findByEdition(
                 CertificationEditionConcept.CERTIFICATION_EDITION_2015.getYear());
         logger.info("Completed retreiving all 2015 listings");
 
-        return listings.parallelStream()
-                .map(dto -> getCertifiedProductSearchDetails(dto.getId(), logger))
+        return listings
+                .subList(0, 99)
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/CertifiedProduct2015Gatherer.java]: Was this subList something for debugging locally?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/onetime/conformancemethodconversion/ConformanceMethodConversionJob.java [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+            }
+        }
+
+        public void deleteAllTestDataForCertificationResult(String chplProductNumber, CertificationResult certificationResult) {
+            for (CertificationResultTestData crtd : certificationResult.getTestDataUsed()) {
+                Query query = entityManager.createQuery(
+                        "SELECT crtde "
+                        + "FROM CertificationResultTestDataEntity crtde "
+                        + "JOIN FETCH crtde.testData tde "
+                        //+ "JOIN FETCH crtde.certificationCriterion cce "
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/onetime/conformancemethodconversion/ConformanceMethodConversionJob.java]: Should this still be here?

**Review by andlar — DISMISSED** [review]

**Review by kekey1 — DISMISSED** [review]

File: chpl/chpl-api/src/main/resources/log4j2-xinclude-file-appenders-json.xml [hunk 5]
```diff
@@ -1039,12 +1039,21 @@
             </Delete>
         </DefaultRolloverStrategy>
     </RollingFile>
+<<<<<<< HEAD
```
Comment (kekey1) [inline chpl/chpl-api/src/main/resources/log4j2-xinclude-file-appenders-json.xml]: This file has a few merge conflicts remaining

**Review by kekey1 — CHANGES_REQUESTED** [review]

**Review by kekey1 — APPROVED** [review]

**Review by andlar — APPROVED** [review]
